gobblin-data-management/src/test/java/org/apache/gobblin/data/management/copy/iceberg/IcebergTableMetadataValidatorUtilsTest.java

```diff
@@ -0,0 +1,200 @@
+      schema1, partitionSpec1, "tableLocationForSchema1WithPartitionSpec1", new HashMap<>());
+  private static final TableMetadata tableMetadataWithSchema3AndUnpartitionedSpec = TableMetadata.newTableMetadata(
+      schema3, unpartitionedPartitionSpec, "tableLocationForSchema3WithUnpartitionedSpec", new HashMap<>());
+  private static final String SCHEMA_MISMATCH_EXCEPTION = "Schema Mismatch between Metadata";
+  private static final String PARTITION_SPEC_MISMATCH_EXCEPTION = "Partition Spec Mismatch between Metadata";
+  private static final boolean VALIDATE_STRICT_PARTITION_EQUALITY_TRUE = Boolean.TRUE;
+  private static final boolean VALIDATE_STRICT_PARTITION_EQUALITY_FALSE = Boolean.FALSE;
+  @Test
+  public void testValidateSameSchema() throws IOException {
+    IcebergTableMetadataValidatorUtils.failUnlessCompatibleStructure(
+        tableMetadataWithSchema1AndUnpartitionedSpec, tableMetadataWithSchema1AndUnpartitionedSpec,
+        VALIDATE_STRICT_PARTITION_EQUALITY_TRUE
+    );
+  }
+
+  @Test
+  public void testValidateDifferentSchema() {
+    // Schema 1 and Schema 2 have different field order
+
+    TableMetadata tableMetadataWithSchema2AndUnpartitionedSpec = TableMetadata.newTableMetadata(schema2IsNotSchema1Compat,
+        unpartitionedPartitionSpec, "tableLocationForSchema2WithUnpartitionedSpec", new HashMap<>());
+
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema1AndUnpartitionedSpec,
+        tableMetadataWithSchema2AndUnpartitionedSpec, SCHEMA_MISMATCH_EXCEPTION);
+  }
+
+  @Test
+  public void testValidateSchemaWithDifferentTypes() {
+    // schema 3 and schema 4 have different field types for field1
+
+    TableMetadata tableMetadataWithSchema4AndUnpartitionedSpec = TableMetadata.newTableMetadata(schema4IsNotSchema3Compat,
+        unpartitionedPartitionSpec, "tableLocationForSchema4WithUnpartitionedSpec", new HashMap<>());
+
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema3AndUnpartitionedSpec,
+        tableMetadataWithSchema4AndUnpartitionedSpec, SCHEMA_MISMATCH_EXCEPTION);
+  }
+
+  @Test
+  public void testValidateSchemaWithEvolvedSchemaI() {
+    // TODO: This test should pass in the future when we support schema evolution
+    // Schema 3 has one more extra field as compared to Schema 1
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema1AndUnpartitionedSpec,
+        tableMetadataWithSchema3AndUnpartitionedSpec, SCHEMA_MISMATCH_EXCEPTION);
+  }
+
+  @Test
+  public void testValidateSchemaWithEvolvedSchemaII() {
+    // TODO: This test should pass in the future when we support schema evolution
+    // Schema 3 has one more extra field as compared to Schema 1
+    verifyFailUnlessCompatibleStructureIOException(tableMetadataWithSchema3AndUnpartitionedSpec,
+        tableMetadataWithSchema1AndUnpartitionedSpec, SCHEMA_MISMATCH_EXCEPTION);
+  }
```

Comment:
looking across these two test cases, when we do later support schema evolution, do we really expect it to be a symmetrical operation (where both cases would pass)?  I'd have thought to only support forward-compatibility, but not backward compat.  i.e. S is compatible w/ S`, but S` is no longer compatible w/ S.

Reply:
That's a valid point, we can discuss about this in detail but totally agree with your point to only support forward-compatibility, but not backward compat.

Removing the comment from first test.